src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,884 @@
+	uint32_t           anchor_index  = 0;
+
+	dfs_obj_t         *obj = NULL;
+	daos_anchor_t      anchor;
+	int                rc = 0;
+
+	struct stat       *stats    = NULL;
+	struct dirent     *dentries = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "LsIOOI", &hdl, &path, &anchor_index, &files, &dirs,
+				       &readdir_chunk);
+
+	assert(hdl->dfs != NULL);
+
+	if (readdir_chunk == 0) {
+		return PyLong_FromLong(-DER_INVAL);
+	}
+
+	D_ALLOC_ARRAY(dentries, readdir_chunk);
+	if (dentries == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	D_ALLOC_ARRAY(stats, readdir_chunk);
+	if (stats == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup object at '%s': %s (rc=%d)", path, strerror(rc), rc);
+		goto out;
+	}
+
+	daos_anchor_init(&anchor, 0);
+	rc = dfs_obj_anchor_set(obj, anchor_index, &anchor);
+	if (rc) {
+		D_ERROR("Could not set anchor '%u' for object at '%s': %s (rc=%d)", anchor_index,
+			path, strerror(rc), rc);
+		goto out;
+	}
+
+	uint32_t nr = readdir_chunk;
+	while (nr > 0) {
+		nr = readdir_chunk;
+		rc = dfs_readdirplus(hdl->dfs, obj, &anchor, &nr, dentries, stats);
+		if (rc) {
+			D_ERROR("Readdirplus of '%s' failed: %s (rc=%d)", path, strerror(rc), rc);
+			goto out;
+		}
+
+		for (uint32_t i = 0; i < nr; ++i) {
+			const struct stat *st    = &stats[i];
+			PyObject          *dname = PyUnicode_FromString(dentries[i].d_name);
+			if (dname == NULL) {
+				rc = ENOMEM;
+				goto out;
+			}
+
+			if (S_ISDIR(st->st_mode)) {
+				rc = PyList_Append(dirs, dname);
+				Py_DECREF(dname); /* PyList_Append does not steal the reference */
+				if (rc) {
+					goto out;
+				}
+				continue;
+			}
+
+			PyObject *item = PyTuple_New(2);
+			if (item == NULL) {
+				rc = ENOMEM;
+				goto out;
+			}
+
+			PyTuple_SetItem(item, 0, dname); /* steals the reference */
+			PyTuple_SetItem(item, 1, PyLong_FromLong(st->st_size));
+
+			rc = PyList_Append(files, item);
+			Py_DECREF(item); /* PyList_Append does not steal the reference */
+			if (rc < 0) {
+				rc = -DER_IO;
+				goto out;
+			}
+		}
+	}
+
+out:
+	D_FREE(dentries);
+	D_FREE(stats);
+
+	if (obj) {
+		dfs_release(obj);
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+static PyObject *
+__shim_handle__torch_read(PyObject *self, PyObject *args)
+{
+	ssize_t            rc     = 0;
+	struct dfs_handle *hdl    = NULL;
+	char              *path   = NULL;
+	dfs_obj_t         *obj    = NULL;
+	PyObject          *buffer = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "LsO", &hdl, &path, &buffer);
+	assert(hdl->dfs != NULL);
+
+	if (!PyObject_CheckBuffer(buffer)) {
+		PyErr_SetString(PyExc_TypeError,
+				"Expected an object that supports the buffer protocol");
+		return NULL;
+	}
+
+	Py_buffer bview;
+	if (PyObject_GetBuffer(buffer, &bview, PyBUF_WRITE) == -1) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not writable");
+		return NULL;
+	}
+
+	if (!PyBuffer_IsContiguous(&bview, 'C')) {
```

Comment:
probably deserve a comment there. Could you please elaborate on what the issue is?